### ISSUE TITLE
Privileges dropped when altering procedure (MariaDB)

### ISSUE THREAD

**Issue by louis77** [HeidiSQL/HeidiSQL#1508]
**Preconditions**

* HeidiSQL version: 11.3..6295
* Database type and version: MariaDB 10.6.5
* OS: Ubuntu 20.04

**Describe the bug**

I clicked on several STORED PROCEDUREs to change parameter names. When I saved the altered procedure, all the EXECUTE privileges for users for these procedures where gone.

**To Reproduce**

This is the SQL that HeidiSQL executed:

```sql
DROP PROCEDURE IF EXISTS `usp_AuthAccessKey`;

CREATE DEFINER=`admin`@`%` PROCEDURE `usp_AuthAccessKey`(
	IN `v_AccessKey` varchar(255)
)
LANGUAGE SQL
NOT DETERMINISTIC
READS SQL DATA
SQL SECURITY DEFINER
COMMENT ''
BEGIN
....
END;
```

Since the original procedure is dropped, the privileges are automatically deleted. HeidiSQL should ALTER the procedure instead of DROPping it.

**Comment by ansgarbecker** [comment]
According to my own comments from b66a62b9f87af4e0ecb8e45383e388a9842c71ea, I think altering a procedure was not possible in older versions of MySQL. [Documentation](https://dev.mysql.com/doc/refman/8.0/en/alter-procedure.html) for MySQL 8 still says one should drop and recreate. Not sure what MariaDB supports though. I can probably implement a different save action for both.

**Comment by louis77** [comment]
MariaDB >= 10.1.3 supports `CREATE OR REPLACE PROCEDURE` and keeps the privileges this way:
https://mariadb.com/kb/en/create-procedure/#or-replace

MySQL 8 uses `ALTER PROCEDURE`:
https://dev.mysql.com/doc/refman/8.0/en/alter-procedure.html

MySQL 5.7 also:
https://dev.mysql.com/doc/refman/5.7/en/alter-procedure.html

I'm not sure what is closer to the SQL standard but this clearly shows that MySQL and MariaDB are two different databases and compatibility will be harder for every now release.

**Comment by ansgarbecker** [comment]
I guess MariaDB has the best working approach then. MySQL 5 + 8 both cannot modify parameters or the body, according to the docs:

> This statement can be used to change the characteristics of a stored procedure. ... However, **you cannot change the parameters or body** of a stored procedure using this statement; to make such changes, you must drop and re-create the procedure using DROP PROCEDURE and CREATE PROCEDURE.

**Comment by louis77** [comment]
Hi @ansgarbecker, today I realized that `ALTER PROCEDURE` indeed only allows the modification of the procedure characteristics like `COMMENT` or `MODIFIES DATA`. It doesn't allow to change parameters or body. Therefore, MariaDB acts like MySQL in a sense that every modification of a procedure leads to a loss of privileges.

I still think that HeidiSQL could check for all privileges granted for a specifc procedure and recreate them after the modification (at least one of the commercial tools I tested does it).
